### ISSUE TITLE
moba_naive.py and moba_efficient.py may not be theoretically equivalent in their computational results.

### ISSUE THREAD

**Issue by yliuls** [MoonshotAI/MoBA#23]
In the moba_naive.py , you consider the cornercases that same topk val should beselected
<img width="543" alt="Image" src="https://github.com/user-attachments/assets/d0f71718-92b3-447b-9fd3-e45729aa43d3" />

But in the moba_efficient.py, you just directly select the topk chunk, do not consider that cornercases.
<img width="484" alt="Image" src="https://github.com/user-attachments/assets/4f720d03-d8a7-41f1-ac04-8036ce6bc1ce" />

**Comment by whitelez** [comment]
Greetings.

* The `moba_naive`  implementation follows these steps:
1. First masks causal relationships by setting gate logits to -inf
2. Then retrieves chunks to be selected based on values
3. Finally uses top_k_idx to select the actual top-k chunks that should be chosen
As you noted, the intermediate step of selecting based on values isn't strictly necessary. A simpler approach using just causal masking with -inf + top_k_idx would be sufficient.

* The `moba_efficient` implementation takes a more direct approach:
1. Directly uses gate -inf masking to select causal chunks
2. Then applies top_k_idx to select chunks

This is an intentional simplification in the `moba_efficient` implementation.
